### DISCUSSION TITLE
Should body.append_operation require a mutable ref?

### DISCUSSION THREAD

**Discussion by jmintb** [mlir-rs/melior#313] [Ideas]
Hi, I was a bit surpised that `module.body().append_operation` does not require a mutable reference. I am not that familiar with MLIR internals yet but the unsafe block inside `append_operation` does appear to be mutating the body.

Can we make this safer by switching to a mutable reference or am I missing something?

**Comment by raviqqe** [comment]
Yes. It's a problem of the current design. I've decided to do so originally following the design of [Inkwell](https://github.com/TheDan64/inkwell) so that we can represent a graph of operations just like a graph at the level of Rust codes.

For example, when we have mutable references instead of immutable ones in those places, we need to write:

```rust
let value = block.append_operation(operation).result(0).unwrap().into();
block.append_operation(func::r#return(&[value], location));
```

rather than:

```rust
block.append_operation(func::r#return(
    &[block.append_operation(operation).result(0).unwrap().into()],
    location,
));
```

because Rust doesn't allow referencing mutable references at two locations at the same time for linearity.

You can see previous notes and discussions about it in those tickets:

- https://github.com/raviqqe/melior/issues/24
- https://github.com/raviqqe/melior/pull/274
- https://github.com/raviqqe/melior/issues/347

As described in #24, we haven't decided the best solution. But if you have any ideas or opinions, we would like to hear that!
